lib/Sema/CSGen.cpp

```diff
@@ -2721,6 +2721,18 @@ namespace {
               locator.withPathElement(LocatorPathElt::PatternMatch(pattern)));
 
           baseType = parentType;
+          // Perform member lookup into the external pattern metatype. e.g.
+          // `case let .test(tuple) as Test`.
+        } else if (externalPatternType) {
+          Type externalMetaType = MetatypeType::get(externalPatternType);
```

Comment:
Hi @xedin ! 
Many thanks for your review!
I am trying to change `else` block like this:

```C++
else {
    if (externalPatternType) {
        CS.addConstraint(
            ConstraintKind::Conversion, patternType, externalPatternType,
            locator.withPathElement(LocatorPathElt::PatternMatch(pattern)));
    }
    // Use the pattern type for member lookup.
    CS.addUnresolvedValueMemberConstraint(
        MetatypeType::get(patternType), enumPattern->getName(),
        memberType, CurDC, functionRefKind,
        locator.withPathElement(LocatorPathElt::PatternMatch(pattern)));

    baseType = patternType;
}
```

It will compile successfully if `patternType` is convertible to `externalPatternType`.
But when `patternType` is not convertible to `externalPatternType` ,
it will throw `error: type of expression is ambiguous without more context`, 
instead of the warning `warning: cast from 'any Bar' to unrelated type 'NonBarBaz' always fails`.

Reply:
I think the current changes of this PR might be enough for is pattern to enum element?

Using `externalPatternType` to add value member constraints to resolve `memberType`.
`patternType` will be connected to cast type via CheckedCast constraint .
And when applying the solution, it will emit the warning at `coercePatternToType` if patternType is not convertible to externalPatternType .

Please feel free to give your comments and thanks for your review!

Reply:
I'm going to think about this a bit more, but let's the testing going meanwhile.

Reply:
Since this is equivalent to TypeCheckPattern behavior, let's land it as is, we can always improve the implementation later.